modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Cursor.scala

```diff
@@ -1,275 +1,243 @@
-
-    def field(p: IriNode => Boolean): Cursor          = field(subject, p, parent)
-    def focus: Option[Node]                           = Some(obj)
-    def values: Option[Iterable[Node]]                = focus.map(List(_))
-    def addOp(cursor: SCursor, op: CursorOp): SCursor = new NodeCursor(subject, obj, parent)(cursor, op, graph)
-    def downField(p: IriNode => Boolean): Cursor      = downField(obj, p)
-    def up: Cursor                                    = parent.addOp(this, MoveUp)
-    override def downSet: Set[Cursor]                 = Set(this)
 
-  }
-
-  private[rdf] final class SetNodeCursorSel(subject: Node, obj: Set[Node], parent: SCursor)(
-      lastCursor: SCursor,
-      lastOp: CursorOp,
-      graph: Graph
-  ) extends FieldCursor(lastCursor, lastOp, graph) {
+  private[this] final case class SetCursor(cursorSet: Set[NodeCursor], lastCursor: Cursor, lastOp: CursorOp, g: Graph)
+      extends Cursor(lastCursor, lastOp, g) {
 
-    def field(p: IriNode => Boolean): Cursor = field(subject, p, parent)
+    override def focus: Option[Node] = None
 
-    override def downSet: Set[Cursor] =
-      obj.map(new NodeCursorArrayElem(subject, _, this)(this, DownSet, graph))
-
-    override def downAt(o: Node => Boolean): Cursor =
-      obj.find(o) match {
-        case None       => fail(DownAt(o))
-        case Some(node) => new NodeCursorArrayElem(subject, node, this)(this, DownAt(o), graph)
+    override def values: Option[Set[Node]] = {
+      val nodes = cursorSet.foldLeft(Set.empty[Node]) {
+        case (set, c) =>
+          c.focus match {
+            case Some(n) => set + n
+            case None    => set
+          }
       }
+      Some(nodes)
+    }
 
-    def focus: Option[Node]                           = None
-    def values: Option[Iterable[Node]]                = Some(obj)
-    def addOp(cursor: SCursor, op: CursorOp): SCursor = new SetNodeCursorSel(subject, obj, parent)(cursor, op, graph)
-    def downField(p: IriNode => Boolean): Cursor      = fail(DownField(p))
-    def up: Cursor                                    = parent.addOp(this, MoveUp)
-
-  }
-
-  private[rdf] final class ListNodeCursorSel(subject: Node, obj: List[Node], parent: SCursor)(
-      lastCursor: SCursor,
-      lastOp: CursorOp,
-      graph: Graph
-  ) extends FieldCursor(lastCursor, lastOp, graph) {
-
-    def field(p: IriNode => Boolean): Cursor = field(subject, p, parent)
-
-    override def downList: List[Cursor] =
-      obj.map(new NodeCursorArrayElem(subject, _, this)(this, DownList, graph))
-
-    override def downSet: Set[Cursor] =
-      downList.toSet
+    override def cursors: Option[Set[Cursor]] =
+      Some(cursorSet.asInstanceOf[Set[Cursor]])
 
-    override def downAt(o: Node => Boolean): Cursor =
-      obj.find(o) match {
-        case None       => fail(DownAt(o))
-        case Some(node) => new NodeCursorArrayElem(subject, node, this)(this, DownAt(o), graph)
+    override def narrow: Cursor =
```

Comment:
lazy val ?

Reply:
I don't see how that would help: we don't expect calling narrow multiple times on the same cursor.